### PR TITLE
Add category section

### PR DESCRIPTION
**Pull request by jyapayne**
This is a follow up to https://github.com/nim-lang/nimforum/pull/194. Sorry @dom96 for another long PR ;)

In general, this PR makes categories useful in that users can filter by them and only see the categories that they care about.

**Features**

- New categories list page, which allow selection of a category, similar to Discourse
- Better display of categories in the category picker dropdown. The dropdown now contains the category description and how many topics are in each category
- Ability to change category when already viewing a category

**Code Changes**

- The "Add category" modal was refactored into it's own file so that it could be added to both the categories list page and the categories dropdown
- A reload was added on Login/Logout because state was not being refreshed in some widgets, most notably the Category picker. There is probably a better way to do this, but not without a good refactor. I'll make a note of this and address it in the future
- Made testing a bit easier by waiting for multiple elements
- Added tests for new categories list page

### CODE REVIEW

File: src/frontend/categorylist.nim [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import options, json, httpcore
```
Comment (jyapayne) [inline src/frontend/categorylist.nim]: This file was refactored from `categorypicker.nim`

File: src/frontend/mainbuttons.nim [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import options
```
Comment (jyapayne) [inline src/frontend/mainbuttons.nim]: This was refactored out of `threadlist.nim`

**Comment by jyapayne** [comment]
And some pictures for your viewing pleasure :)

Screen Shot 2020-02-24 at 6 54 59 PM
Screen Shot 2020-02-24 at 6 55 21 PM
Screen Shot 2020-02-24 at 6 55 40 PM
Screen Shot 2020-02-24 at 6 55 54 PM
Screen Shot 2020-02-24 at 6 56 12 PM

**Comment by jyapayne** [comment]
@Clyybber @Araq @dom96 reminder ping! Feel free to tear it apart. There were some struggles with elements not refreshing (which is why I added reloading on signup/login/logout), which could probably be done in a different way (advice would be nice here).

Also wanted to add that while this PR looks fairly big, lots of changes are refactoring and tests.

File: src/frontend/header.nim [hunk 3]
```diff
@@ -31,44 +32,49 @@ when defined(js):
...
         () => state.signupModal.show()
       ),
       signupModal: newSignupModal(
-        () => (state.lastUpdate = fromUnix(0); getStatus()),
+        () => (state.lastUpdate = fromUnix(0); getStatus(reload=true)),
         () => state.loginModal.show()
       ),
       userMenu: newUserMenu(
-        () => (state.lastUpdate = fromUnix(0); getStatus(logout=true))
+        () => (state.lastUpdate = fromUnix(0); getStatus(logout=true, reload=true))
```
Comment (dom96) [inline src/frontend/header.nim]: Why was this necessary? In general we should be reloading less not more, and reloading the page when logged in breaks the SPA illusion quite heavily.

File: src/frontend/threadlist.nim [hunk 4]
```diff
@@ -239,10 +222,11 @@ when defined(js):
                   td(colspan="6"):
                     tdiv(class="loading loading-lg")
                 else:
-                  td(colspan="6", onClick=onLoadMore):
+                  td(colspan="6",
+                     onClick = (ev: Event, n: VNode) => (onLoadMore(ev, n, categoryId))):
                     span(text "load more threads")
 
-  proc renderThreadList*(currentUser: Option[User]): VNode =
+  proc renderThreadList*(currentUser: Option[User], categoryId = -1): VNode =
```
Comment (dom96) [inline src/frontend/threadlist.nim]: This should be an `Option`, avoid using magical constants like `-1`.

File: src/frontend/header.nim (hunk 3 above)
Comment (jyapayne) [reply src/frontend/header.nim]: I was trying to find a way to not do this, and I agree it breaks the SPA illusion. I did want some feedback on it but Araq merged the PR :o. Ideally there would be some way to refresh the entire Karax state or some way to tell specific components to reload state. In particular, the category picker would not get it's state reloaded on logout, which would leave the add category functionality available on logout from an admin account. It results in an error when trying to submit, but that button should be hidden and the component should reload. Just wasn't sure how to do that in Karax. A redraw call didn't help

File: src/frontend/threadlist.nim (hunk 4 above)
Comment (jyapayne) [reply src/frontend/threadlist.nim]: Sure, I can do that.

File: src/frontend/header.nim (hunk 3 above)
Comment (Araq) [reply src/frontend/header.nim]: Actually I did see it in my review but I thought "I'm sure he has his reasons". If you mind how this worked out, exclude me from nimforum's development process or give @jyapayne write access. Encouraging contributions is more important to me than striving for perfection, it's a forum software after all.

**Comment by jyapayne** [comment]
@dom96 I addressed your issues in https://github.com/nim-lang/nimforum/pull/232 and https://github.com/nim-lang/nimforum/pull/231

File: src/frontend/header.nim (hunk 3 above)
Comment (dom96) [reply src/frontend/header.nim]: At the very least write "Looked through it, looks good" instead of silently merging. We won't know whether you've reviewed it or not otherwise.
